src/System.Private.CoreLib/src/System/GC.cs

```diff
@@ -508,5 +513,66 @@ public static void EndNoGCRegion()
         {
             EndNoGCRegionWorker();
         }
+
+        // generation: -1 means to let GC decide (equivalent to regular new T[])
+        // 0 means to allocate in gen0
+        // GC.MaxGeneration means to allocate in the oldest generation
+        //
+        // pinned: true means you want to pin this object that you are allocating
+        // otherwise it's not pinned.
+        //
+        // alignment: only supported if pinned is true.
+        // -1 means you don't care which means it'll use the default alignment.
+        // otherwise specify a power of 2 value that's >= pointer size
+        // the beginning of the payload of the object (&array[0]) will be aligned with this alignment.
+        // static T[] AllocateArray<T>(int length, int generation = -1, bool pinned = false, int alignment = -1)
+        public static T[] AllocateArray<T>(int length, int generation = -1, bool pinned = false, int alignment = -1)
```

Comment:
If we're not going to support the additional arguments, we shouldn't have the additional arguments.  We could add those in an overload if/when we're able to support them.

Regardless, the API should go through API review before this gets checked in as public.  If we want to check it in as internal for now for use within corelib, e.g. in ArrayPool, sounds good to me.

Reply:
I kind of feel the same. Just having `AllocateUninitializedArray<T>(size)` might be sufficient for now with more overloads to follow. But I could be convinced either way. 
@terrajobst   - any input on this?

Reply:
Having an ability to control array allocation alignment is important for efficient HW intrinsic use. Since API is not approved yet why don't we have internal implementation with alignment. IMHO it will help with optimization in several places in BCL.

Reply:
yes, having alignment support is certainly useful. we do have it implemented but not tested and testing would require work obviously and will not make 3.0. but we will look into enabling it in a future release.

Reply:
@terrajobst could give us a timeslot in your API review meeting? when is the next one?

Reply:
>  will not make 3.0

Is any of this required for 3.0? I think this whole API should be post 3.0.

Reply:
we would like to get the part of the API that does the uninit in 3.0. it takes a lot of work to implement everything that the AllocateArray API support and I don't see why we need to wait till everything is implemented before publishing the API.

Reply:
> we would like to get the part of the API that does the uninit in 3.0

We are trying to lock down 3.0. Unless there is somebody blocked by this API not being available in 3.0, we should look at doing the whole thing for after 3.0.